// react/lib/ReactServerRenderingTransaction.js
define('node_modules/react/lib/ReactServerRenderingTransaction', function(require, exports, module) {

  /**
   * Copyright 2014-present, Facebook, Inc.
   * All rights reserved.
   *
   * This source code is licensed under the BSD-style license found in the
   * LICENSE file in the root directory of this source tree. An additional grant
   * of patent rights can be found in the PATENTS file in the same directory.
   *
   * @providesModule ReactServerRenderingTransaction
   */
  
  'use strict';
  
  var _assign = require('node_modules/object-assign/index');
  
  var PooledClass = require('node_modules/react/lib/PooledClass');
  var Transaction = require('node_modules/react/lib/Transaction');
  var ReactInstrumentation = require('node_modules/react/lib/ReactInstrumentation');
  var ReactServerUpdateQueue = require('node_modules/react/lib/ReactServerUpdateQueue');
  
  /**
   * Executed within the scope of the `Transaction` instance. Consider these as
   * being member methods, but with an implied ordering while being isolated from
   * each other.
   */
  var TRANSACTION_WRAPPERS = [];
  
  if ('development' !== 'production') {
    TRANSACTION_WRAPPERS.push({
      initialize: ReactInstrumentation.debugTool.onBeginFlush,
      close: ReactInstrumentation.debugTool.onEndFlush
    });
  }
  
  var noopCallbackQueue = {
    enqueue: function enqueue() {}
  };
  
  /**
   * @class ReactServerRenderingTransaction
   * @param {boolean} renderToStaticMarkup
   */
  function ReactServerRenderingTransaction(renderToStaticMarkup) {
    this.reinitializeTransaction();
    this.renderToStaticMarkup = renderToStaticMarkup;
    this.useCreateElement = false;
    this.updateQueue = new ReactServerUpdateQueue(this);
  }
  
  var Mixin = {
    /**
     * @see Transaction
     * @abstract
     * @final
     * @return {array} Empty list of operation wrap procedures.
     */
    getTransactionWrappers: function getTransactionWrappers() {
      return TRANSACTION_WRAPPERS;
    },
  
    /**
     * @return {object} The queue to collect `onDOMReady` callbacks with.
     */
    getReactMountReady: function getReactMountReady() {
      return noopCallbackQueue;
    },
  
    /**
     * @return {object} The queue to collect React async events.
     */
    getUpdateQueue: function getUpdateQueue() {
      return this.updateQueue;
    },
  
    /**
     * `PooledClass` looks for this, and will invoke this before allowing this
     * instance to be reused.
     */
    destructor: function destructor() {},
  
    checkpoint: function checkpoint() {},
  
    rollback: function rollback() {}
  };
  
  _assign(ReactServerRenderingTransaction.prototype, Transaction.Mixin, Mixin);
  
  PooledClass.addPoolingTo(ReactServerRenderingTransaction);
  
  module.exports = ReactServerRenderingTransaction;

});
